core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+					x_thumbnail := cast(int)compress.read_u8(ctx) or_return
+					y_thumbnail := cast(int)compress.read_u8(ctx) or_return
+					thumbnail: []image.RGB_Pixel
+
+					if x_thumbnail * y_thumbnail != 0 {
+						greyscale_thumbnail := false
+						thumbnail_size := x_thumbnail * y_thumbnail * 3
+						// According to the JFIF spec, the thumbnail should always be made of RGB pixels.
+						// But some jpegs encode single-channel thumbnails.
+						if thumbnail_size != length - 14 && thumbnail_size / 3 == length - 14 {
+							thumbnail_size = x_thumbnail * y_thumbnail
+							greyscale_thumbnail = true
+						} else {
+							return img, .Invalid_Thumbnail_Size
+						}
+						thumb_pixels := slice.reinterpret([]image.RGB_Pixel, compress.read_slice_from_memory(ctx, x_thumbnail * y_thumbnail) or_return)
+
+						if .return_metadata in options {
+							thumbnail = make([]image.RGB_Pixel, x_thumbnail * y_thumbnail)
+							copy(thumbnail, thumb_pixels)
+
+							info: ^image.JPEG_Info
+							if img.metadata == nil {
+								info = new(image.JPEG_Info)
+							} else {
+								info = img.metadata.(^image.JPEG_Info)
+							}
+							info.jfif_app0 = image.JFIF_APP0{
+								version,
+								units,
+								x_density,
+								y_density,
+								cast(u8)x_thumbnail,
+								cast(u8)y_thumbnail,
+								thumbnail,
+								greyscale_thumbnail,
+							}
+							img.metadata = info
+						}
+					}
+				} else if slice.equal(ident[:], image.JFXX_Magic[:]) {
+					extension_code := cast(image.JFXX_Extension_Code)compress.read_u8(ctx) or_return
+					thumbnail: []byte
+
+					switch extension_code {
+					// We return the JPEG-compressed bytes for this type of thumbnail.
+					// It's up to the user if they want to decode it by checking the extension code
+					// and calling image.load() on the thumbnail.
+					// Not sure where to document that though, maybe it's better if the thumbnail is always raw pixel data.
+					case .Thumbnail_JPEG:
+						// +1 for the NUL byte
+						thumbnail_len := length - (size_of(image.JFXX_Magic) + 1 + size_of(image.JFXX_Extension_Code))
+						thumbnail_jpeg := compress.read_slice(ctx, thumbnail_len) or_return
+
+						if .return_metadata in options {
+							thumbnail = make([]byte, thumbnail_len)
+							copy(thumbnail, thumbnail_jpeg)
+
+							info: ^image.JPEG_Info
+							if img.metadata == nil {
+								info = new(image.JPEG_Info)
+							} else {
+								info = img.metadata.(^image.JPEG_Info)
+							}
+							info.jfxx_app0 = image.JFXX_APP0{
+								extension_code,
+								0,
+								0,
+								thumbnail,
+							}
+							img.metadata = info
+						}
+					case .Thumbnail_3_Byte_RGB:
+						x_thumbnail := cast(int)compress.read_u8(ctx) or_return
+						y_thumbnail := cast(int)compress.read_u8(ctx) or_return
+						pixels := compress.read_slice(ctx, x_thumbnail * y_thumbnail * 3) or_return
+
+						if .return_metadata in options {
+							thumbnail = make([]byte, x_thumbnail * y_thumbnail * 3)
```

Comment:
Can run out of memory, should handle the optional error from `make` here and return it as an error.